Add more DataFlow tests from linker to NativeAOT

vitek-karas opened the pull request:
This adds several dataflow tests from linker - only those which are passing with the fixes in this change. More will be added later.

Fixes:
- Correctly handle Requires attributes in generic types
- Recognize `ref Type` as interesting for data flow
- Fix generic parameter propagation for compiler generated types
- Treat multi-dimensional arrays as "Unknown" values (to match linker behavior)

vitek-karas committed:
Add more DataFlow tests from linker to NativeAOT

This adds several dataflow tests from linker - only those which are passing with the fixes in this change. More will be added later.

Fixes:
- Correctly handle Requires attributes in generic types
- Recognize `ref Type` as interesting for data flow
- Fix generic parameter propagation for compiler generated types
- Treat multi-dimensional arrays as "Unknown" values (to match linker behavior)

vitek-karas commented:
The changes to the tests are ported back to linker in https://github.com/dotnet/linker/pull/2912.

vitek-karas committed:
Test updates from linker

tlakollo commented on src/coreclr/tools/aot/Mono.Linker.Tests.Cases/DataFlow/MethodByRefReturnDataFlow.cs:
```diff
@@ -0,0 +1,85 @@
+
+		// Correct behavior in the linker, but needs to be added in analyzer
+		// Bug link: https://github.com/dotnet/linker/issues/2158
+		[ExpectedWarning ("IL2026", "Message for --TestType.Requires--", ProducedBy = ProducedBy.Trimmer)]
```
Should we have a similar issue in NativeAOT?

vitek-karas replied:
I will look into this - I should understand why it's not producing the warning (linker/AOT should behave the same in this case as it's basically the same code).

vitek-karas replied:
This has been solved in later commits - it's now enabled for NativeAOT as well.

tlakollo commented on src/coreclr/tools/aot/Mono.Linker.Tests.Cases/DataFlow/UnresolvedMembers.cs:
```diff
@@ -0,0 +1,135 @@
+		{ }
+
+		[Kept]
+		[ExpectedWarning ("IL2066", "TypeWithUnresolvedGenericArgument", ProducedBy = ProducedBy.Trimmer | ProducedBy.Analyzer)] // Local variable type
```
Is there an issue to track this in NativeAOT?

vitek-karas replied:
See the comment on top of this test file. AOT will actually replace the method bodies with throw and thus won't produce any warnings really. And it's by design - the method would throw when executed (JIT could not compile it), so there's no point in producing warnings for it in AOT.

Technically linker could do the same, but the current behavior of linker is to "ignore but keep" unresolvable references.

tlakollo reviewed (commented):
Are we sure about having all these tests whose main focus is to verify if things are kept? I think once we start to validate the kept attribute now we will have to do it 100% right or we will break CI

vitek-karas committed:
Merge remote-tracking branch 'dotnet/main' into AotTests2

vitek-karas committed:
Improve attribute data flow by providing the correct MessageOrigin. This changes MetadataManager interface, but it should not be a problem.

This fixes the attribute test failures.

The CompilerGeneratedCodeDataflow test was failing because AOT didn't scan the body of the MoveNext method in the state machine for iterators. This is because the test actually never references IEnumerator<>.MoveNext and so the interface method and all of its implementations are removed. Fixed by making an explicit reference to it at least once.

vitek-karas commented:
Please see https://github.com/dotnet/runtime/pull/72777/commits/c1ca4bc3478d6f18aaf17db78e80f35a306296db for fixes to all CI failures.

vitek-karas committed:
Fix a bug in ref-return-value data flow

This was missed during porting of the changes from linker.

More cleanup of tests.

vitek-karas committed:
Fix attribute parent calculation by forcing it to be passed from the caller

vitek-karas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/CustomAttributeBasedDependencyAlgorithm.cs:
```diff
@@ -77,32 +77,32 @@ public static void AddDependenciesDueToCustomAttributes(ref DependencyList depen
         }
 
-        private static void AddDependenciesDueToCustomAttributes(ref DependencyList dependencies, NodeFactory factory, EcmaModule module, CustomAttributeHandleCollection attributeHandles)
+        private static void AddDependenciesDueToCustomAttributes(ref DependencyList dependencies, NodeFactory factory, EcmaModule module, CustomAttributeHandleCollection attributeHandles, TypeSystemEntity parent)
```
@MichalStrehovsky could you please take a look at the changes around attributes - passing the "parent" around.
Originally I tried using the CustomAttribute.Parent handle, but the EcmaModule can't translate lot of handles to type system entities (assembly, generic parameters, method parameters, ...) so it seemed easier to force the decision onto the caller.

MichalStrehovsky replied:
Yep, this works. We don't intern them at GetObject level because these are not interesting entities to resolve for the type system.

vitek-karas commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/DependencyAnalysis/CustomAttributeBasedDependencyAlgorithm.cs:
```diff
@@ -26,20 +26,20 @@ public static void AddDependenciesDueToCustomAttributes(ref DependencyList depen
             {
                 GenericParameter parameter = reader.GetGenericParameter(genericParameterHandle);
-                AddDependenciesDueToCustomAttributes(ref dependencies, factory, method.Module, parameter.GetCustomAttributes());
+                AddDependenciesDueToCustomAttributes(ref dependencies, factory, method.Module, parameter.GetCustomAttributes(), method);
```
We could probably get a `GenericParameterDesc` here, but it seemed relatively complicated to do (would have to enumerate all of them on methoddesc and match it to the one I'm looking for). It probably makes very little difference, since symbols would not be able to map it more precisely to a location anyway.

MichalStrehovsky commented on src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/MethodBodyScanner.cs:
```diff
@@ -1237,6 +1226,14 @@ internal MultiValue DereferenceValue(MultiValue maybeReferenceValue, ValueBasicB
             ValueNodeList methodArguments = PopCallArguments(currentStack, calledMethod, callingMethodBody, isNewObj,
                                                              offset, out newObjValue);
 
+            // Multi-dimensional array access is represented as a call to a special Get method on the array (runtime provided method)
+            // We don't track multi-dimensional arrays in any way, so return unknown value.
+            if (calledMethod is ArrayMethod { Kind: ArrayMethodKind.Get })
+            {
+                currentStack.Push(new StackSlot(UnknownValue.Instance));
+                return;
+            }
```
What problem is this fixing?

There are other kinds of magic methods that don't cast to EcmaMethod (if that's the problem). I would rather just have this bail when we need to cast (if that's what the problem is).

When I was originally doing the first DataFlow iteration, I ran it once in mode where we would run the dataflow scanner on all IL we compile just to make sure it can handle all the patterns (i.e. make it so that everything is interesting to dataflow and do a testpass either in CI or locally).

vitek-karas replied:
It's not about the method being EcmaMethod - the code worked fine without this change. But it was providing different behavior as compared to linker/analyzer. Linker/analyzer doesn't see multi-dimensional array access as a method call, so the difference is when trying to get a value from multi-dimensional array:
* linker/analyzer -> Unknown value - since it's not a handled case
* AOT -> return value from a magical method

The warnings around this are "Unknown value passed into ..." and "<magical method> return value passed into ..." respectively for AOT.

If we think this is OK to have a different I'd be fine removing this and fixing the tests.

MichalStrehovsky replied:
In that case keeping it sounds like a good plan. We might want to extend it to the `Address` method.

The repro for that is:

```csharp
Type[,] types = new Type[1, 1];
ref Type t = ref types[0, 0];
t?.GetMethods();
```

For a moment I thought I would ask how Cecil surfaces it, but maybe I don't want to know. I probably knew in the past but now it's a repressed memory (it is a method call in IL).

vitek-karas replied:
The Cecil's representation is actually not "bad" - it is the method call, but it's unresolvable method, so we fail to process the method call and return Unknown. In NativeAOT the typesystem actually resolves the method... that is the difference.

vitek-karas replied:
The Address works by accident - it uses `ldind_ref` instruction which we don't handle so it's going to end up `Unknown` regardless of how we handle the method call. I will make the change anyway.

vitek-karas committed:
Handle multi-dimensional array Address method

MichalStrehovsky reviewed (approved):
Thank you!
